packages/xod-project/src/patch.js

```diff
@@ -934,23 +934,27 @@ export const upsertDeadPins = def(
   (node, currentPatch, pins) => {
     const nodeId = Node.getNodeId(node);
     const pinKeys = R.keys(pins);
-    const pinsByDir = R.applySpec({
-      [CONST.PIN_DIRECTION.INPUT]: R.filter(Pin.isInputPin),
-      [CONST.PIN_DIRECTION.OUTPUT]: R.filter(Pin.isOutputPin),
-    })(R.values(pins));
+
+    const pinsByDir = R.compose(R.groupBy(Pin.getPinDirection), R.values)(pins);
 
     const rejectNondeadLinks = R.reject(
       R.either(
-        R.compose(isAmong(pinKeys), Link.getLinkInputPinKey),
-        R.compose(isAmong(pinKeys), Link.getLinkOutputPinKey)
+        R.both(
+          R.pipe(Link.getLinkInputNodeId, R.equals(nodeId)),
```

Comment:
Could be replaced with `Link.isLinkInputNodeIdEquals(nodeId)` and the same for output...

Reply:
Oh, right. Done!